wire-compiler/src/main/java/com/squareup/wire/model/Linker.java

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright (C) 2015 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.wire.model;
+
+import com.squareup.protoparser.DataType;
+import com.squareup.protoparser.ProtoFile;
+import com.squareup.wire.IO;
+import java.io.File;
+import java.io.IOException;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public final class Linker {
+  private final String repoPath;
+  private final IO io;
+  private final Set<String> loadedDependencies;
+  private final Map<String, WireType> protoTypeNames;
+  private final Map<ProtoFile, WireProtoFile> protoFilesMap;
+
+  // Context when linking.
+  private final String enclosingProtoPackage;
+  private final List<WireType> enclosingTypes;
+
+  public Linker(String repoPath, IO io) {
+    this.repoPath = repoPath;
+    this.io = io;
+    this.loadedDependencies = new LinkedHashSet<String>();
+    this.protoTypeNames = new LinkedHashMap<String, WireType>();
+    this.protoFilesMap = new LinkedHashMap<ProtoFile, WireProtoFile>();
+    this.enclosingProtoPackage = null;
+    this.enclosingTypes = Collections.emptyList();
+  }
+
+  private Linker(Linker enclosing, String protoPackage, WireType type) {
+    this.repoPath = enclosing.repoPath;
+    this.io = enclosing.io;
+    this.loadedDependencies = enclosing.loadedDependencies;
+    this.protoTypeNames = enclosing.protoTypeNames;
+    this.protoFilesMap = enclosing.protoFilesMap;
+    this.enclosingProtoPackage = protoPackage;
+    this.enclosingTypes = type != null
+        ? Util.concatenate(enclosing.enclosingTypes, type)
+        : enclosing.enclosingTypes;
+  }
+
+  /** Recursively add {@code protoFile} and its dependencies. */
+  public void add(ProtoFile protoFile) throws IOException {
+    WireProtoFile wireProtoFile = new WireProtoFile(protoFile);
+    protoFilesMap.put(protoFile, wireProtoFile);
+
+    // Register the enclosed types.
+    wireProtoFile.register(this);
```

Comment:
Registration is a simple operation. Did you consider doing the PP model walk here rather than in the Wire model to build registered types? Then you could factory the Wire model with the registry completely as immutable objects?